### DISCUSSION TITLE
Can not build aur/opensnitch(-git) on Manjaro

### DISCUSSION THREAD

**Discussion by 0xphk** [evilsocket/opensnitch#768] [General]
Hi, 
I was trying to update opensnitch (aur/opensnitch-git) on Manjaro but the build always fails.
I used it for almost one year and updated it successfully in the past multiple times, but now it won't build any more.
I've removed the opensnitch-git version and tried the aur/opensnitch version instead but it leads to the same error.

I get the following errors and don't know how to address this as I'm not familiar with the go build system.
```
==> Making package: opensnitch 1.5.2-1 (Fr 11 Nov 2022 22:58:47 CET)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> WARNING: Using existing $srcdir/ tree
==> Starting build()...
go: downloading github.com/golang/protobuf v1.5.2
go: downloading google.golang.org/protobuf v1.26.0
# github.com/golang/protobuf/protoc-gen-go
loadinternal: cannot find runtime/cgo
go: downloading google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.2.0
go: downloading google.golang.org/grpc v1.50.1
go: downloading google.golang.org/protobuf v1.27.1
# google.golang.org/grpc/cmd/protoc-gen-go-grpc
loadinternal: cannot find runtime/cgo
protoc-gen-go: /home/phk/.cache/yay/opensnitch/src/gopath/bin/protoc-gen-go
protoc-gen-go-grpc: /home/phk/.cache/yay/opensnitch/src/gopath/bin/protoc-gen-go-grpc
~/.cache/yay/opensnitch/src/opensnitch-1.5.2/proto ~/.cache/yay/opensnitch/src/opensnitch-1.5.2
protoc -I. ui.proto --go_out=../daemon/ui/protocol/ --go-grpc_out=../daemon/ui/protocol/ --go_opt=paths=source_relative --go-grpc_opt=paths=source_relative
python3 -m grpc_tools.protoc -I. --python_out=../ui/opensnitch/ --grpc_python_out=../ui/opensnitch/ ui.proto
~/.cache/yay/opensnitch/src/opensnitch-1.5.2
~/.cache/yay/opensnitch/src/opensnitch-1.5.2/daemon ~/.cache/yay/opensnitch/src/opensnitch-1.5.2
go: downloading github.com/google/gopacket v1.1.14
go: downloading github.com/vishvananda/netlink v0.0.0-20210811191823-e1a867c6b452
go: downloading github.com/fsnotify/fsnotify v1.4.7
go: downloading golang.org/x/net v0.0.0-20191028085509-fe3aa8a45271
go: downloading google.golang.org/grpc v1.32.0
go: downloading github.com/iovisor/gobpf v0.2.0
go: downloading github.com/evilsocket/ftrace v1.2.0
go: downloading github.com/google/nftables v0.0.0-20210514154851-a285acebcad3
go: downloading golang.org/x/sys v0.0.0-20200728102440-3e129f6d46b1
go: downloading github.com/vishvananda/netns v0.0.0-20200728191858-db3c7e526aae
go: downloading google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
go: downloading github.com/mdlayher/netlink v0.0.0-20191009155606-de872b0d824b
go: downloading github.com/koneu/natend v0.0.0-20150829182554-ec0926ea948d
go: downloading github.com/golang/protobuf v1.5.0
go: downloading golang.org/x/text v0.3.0
# github.com/evilsocket/opensnitch/daemon/netfilter
cgo-gcc-prolog: In function ‘_cgo_aceefaf8538a_Cfunc_stop_reading_packets’:
cgo-gcc-prolog:451:49: warning: unused variable ‘_cgo_a’ [-Wunused-variable]
# github.com/iovisor/gobpf/elf
cgo: cannot load DWARF output from $WORK/b084//_cgo_.o: zlib: invalid header
make: *** [Makefile:16: opensnitchd] Error 2
==> ERROR: A failure occurred in build().
    Aborting...
 -> error making: opensnitch
```
I think I had a similar error in the past, but unfortunately can not remember what the solution actually was.
Any help is highly appreciated.

Thanks in advance.

**Comment by gustavo-iniguez-goya** [comment]
hey @0xphk , honestly, no idea why it's failing. It looks like something is not right with the Go installation?
`loadinternal: cannot find runtime/cgo` and `zlib: invalid header`

what's your `$ go env` output?

**Reply by 0xphk** [reply]
Go environment:
```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/phk/.cache/go-build"
GOENV="/home/phk/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/phk/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/phk/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.19.2"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3656454699=/tmp/go-build -gno-record-gcc-switches"
```
I tested CGO_ENABLED=0 after a little bit of searching as I wasn't sure if this is a cross build but it gave the same error.
The discussion in the AUR package page lists the same issue, so maybe it's affecting more people.
https://aur.archlinux.org/packages/opensnitch#comment-881733

**Reply by gustavo-iniguez-goya** [reply]
interesting. I've updated my manjaro installation and the `cannot find runtime/cgo` error appears, but it compiled fine.
Anyway, I've git cloned the repo from github, compiled it manually and it hasn't failed.

My go version is 1.17.5 (the system is outdated).

**Reply by 0xphk** [reply]
I'm pretty sure that I had this message on older Go versions as well but it could compile and update successfully. I suspect it has something to do with the latest Go 1.19 version. I'm currently installing a fresh Manjaro VM to see if the build error is reproducible. Will report back.

Can you please share some details on how to manually build it? I only have little to no Go build process knowledge but would like to give it a try. Thanks

**Reply by gustavo-iniguez-goya** [reply]
see here: https://github.com/evilsocket/opensnitch/wiki/Compilation

By the way, I've updated my manjaro installation and I've reinstalled the opensnitch-git AUR package (now with golang 1.19.3, ) and finished successfully.

**Reply by 0xphk** [reply]
Hopefully it's fixed in the newer version, I'm still on ```go version go1.19.2 linux/amd64```, will update later after VM test is done.

**Reply by 0xphk** [reply]
I have another question, I noticed that I need root privileges when I try to remove the cloned repo in my standard user account in ~/.cache/yay/opensnitch(-git).
There are a lot of directories in this path with permissions ```0555 dr-xr-xr-x``` or even ```0111 d--x--x--x``` which is very weird and I'm sure this isn't intended, or is it? 

This leads also to be unable to reinstall or cleanbuild the package as it throws tons of permission denied errors.
Can somebody please check that the perms are wrong too? This is the same on both my running system and a newly setup vm.
```~/.cache/yay/opensnitch/ $ sudo find -type d -not -perm 0755``` or
```~/.cache/yay/opensnitch/ $ sudo find -type d -perm 0555``` or
```~/.cache/yay/opensnitch/ $ sudo find -type d -perm 0111```

**Reply by 0xphk** [reply]
VM did end up with the same error on Go 1.19.3 and unable to build.

Did some more testing and can now confirm that building works if I use the latest Go 1.18 version 
```go version go1.18.8 linux/amd64```.

The ```loadinternal: cannot find runtime/cgo``` message is still there but the 
```cgo: cannot load DWARF output from $WORK/b084//_cgo_.o: zlib: invalid header``` error isn't shown and compiling works.

So far a workaround for me is to extract the Go 1.18 release somewhere and put it in ```$PATH``` before the newer/systemwide ```go``` binary location.

**Reply by 0xphk** [reply]
Can you please take a look at the directory permissions?

**Reply by gustavo-iniguez-goya** [reply]
> There are a lot of directories in this path with permissions 0555 dr-xr-xr-x or even 0111 d--x--x--x

```bash
~ $ sudo find .cache/yay/opensnitch -type d -perm 0555
~ $ sudo find .cache/yay/opensnitch -type d -not -perm 0755
~ $ sudo find .cache/yay/opensnitch -type d -perm 0111
~ $
```

everything seems normal here.

**Reply by 0xphk** [reply]
That's strange, could confirm this on 3 different Manjaro systems but only for ```aur/opensnitch-git```,
I switched to ```aur/opensnitch``` and there is not problem so I stick to this.

EDIT: With the now working Go 1.18.8 build, the permission problems are also gone, so I expect it must have had something to do with the interrupted build process, leaving some directories in some strange state.

In any case, thanks for the help.
